File: Algorithms/快乐数.java
public class Solution {
    public boolean isHappy(int n) {
		boolean r = false;
		int x=n,y=n;
		while(n > 1){
			x = cal(x);
			if(x == 1) return true;
			y = cal(cal(y));
			if(y == 1) return true;
			if(y == x) return false;
		}
		return true; 
    }
	public int cal(int n){
		int temp = 0;
		while(n != 0){
			temp += Math.pow(n%10,2);
			n /= 10;
		}
		return temp;
	}
}

File: Algorithms/快乐数.py
class Solution:
    def isHappy(self, n: int) -> bool:
        has_vis = set()
        has_vis.add(n)
        while True:
            n = sum([int(_) * int(_) for _ in str(n)])
            if n == 1:
                return True
            if n in has_vis:
                return False
            has_vis.add(n)